benchexec/tools/metaval.py

```diff
@@ -108,14 +111,22 @@ def cmdline(self, executable, options, tasks, propertyfile=None, rlimits={}):
                 ).Tool()
 
         if verifierName in self.wrappedTools:
-            with self._in_tool_directory(verifierName) as oldcwd:
-                wrappedOptions = self.wrappedTools[verifierName].cmdline(
-                    self.wrappedTools[verifierName].executable(),
-                    options,
-                    [os.path.relpath(os.path.join(oldcwd, "output/ARG.c"))],
-                    os.path.relpath(os.path.join(oldcwd, propertyfile)),
-                    rlimits,
+            tool = self.wrappedTools[verifierName]
+            assert isinstance(
+                tool, BaseTool2
+            ), "we expect that all wrapped tools extend BaseTool2"
+            wrapped_executable = self.wrappedTools[verifierName].executable(
```

Comment:
I wonder whether it makes sense to also call `tool.version()` afterwards. The BenchExec docs do not strictly guarantee that this method is called, but they do mention that the general methods (which includes `version()`) are called before the run-specific methods, and it specifically allows tool-info modules to store the result from e.g. `version()` in a field.¹

The reason why this is relevant is that some tool-info modules store information about the tool version to support different versions. I think most do this in the call to `executable()`, but some might do it in `version()`. So to make sure, we could call the latter method here.

And to make it efficient (a call to `version()` typically starts the verifier) we could move both calls into the block that loads the wrapped tool-info module, so that they are called only once.

¹ The commit on which this PR is based is so old that it does not include this part of the docs, in case you wonder where it is. Merging current master into this PR might make sense.

Reply:
I am totally lost here and cannot tell what you want me to do.

> we could move both calls into the block that loads the wrapped tool-info module, so that they are called only once.

How will this lead to the tools only being executed once? Which second call apart from the one cause by `version()` are you referring to? What do you mean with `loads the wrapped tool-info module`? Do you mean the import and instantiation of the wrapped verifiers `Tool` class?

> The BenchExec docs do not strictly guarantee that this method is called, but they do mention that the general methods (which includes version()) are called before the run-specific methods, and it specifically allows tool-info modules to store the result from e.g. version() in a field

If there is no strict guarantee that these methods are called, then this should be formalized. Tools really need to work even if their `version()` method is not called. Everything else would be bad software architecture.

Reply:
> > we could move both calls into the block that loads the wrapped tool-info module, so that they are called only once.
> 
> How will this lead to the tools only being executed once?

The code block that loads the wrapped tool-info module (inside `if verifierName not in self.wrappedTools:`) is executed only once. If `tool.version()` is called only there, it will also be called only once.

> Which second call apart from the one cause by `version()` are you referring to?

With "both" I meant the calls to `executable()` and `version()` of the wrapped module.

> What do you mean with `loads the wrapped tool-info module`? Do you mean the import and instantiation of the wrapped verifiers `Tool` class?

Exactly.

> > The BenchExec docs do not strictly guarantee that this method is called, but they do mention that the general methods (which includes version()) are called before the run-specific methods, and it specifically allows tool-info modules to store the result from e.g. version() in a field
> 
> If there is no strict guarantee that these methods are called, then this should be formalized. Tools really need to work even if their `version()` method is not called. Everything else would be bad software architecture.

Note that this is not about tools not working if their version is not retrieved, but about tool-info modules not working without `version()` being called. And because many tool-info modules have methods whose implementation depends on the result of `version()` but this method is somewhat costly, we consider it ok to cache its result as a side effect in a field (maybe not the best architecture, but not bad, and we do not want to require tool-info modules to be perfect).

If we are going to formalize this, I would tend to say that we will guarantee that `executable()` and `version()` will be called, because this would make writing tool-info modules easier with little to no cost for BenchExec itself, and there are many tool-info modules but only very few callers. Furthermore, this way we do not introduce a new requirement for tool-info modules right after the new API was introduced.

Of course, MetaVal's tool-info module needs to support only a limited set of other tool-info modules and for these it might be fine if `version()` is never called (it is fine for CPAchecker's module, for example). So you can also check whether this is the case for all tool-info modules that you want to support and keep the current code (though a comment about this would probably be valuable for the future).

Reply:
> Of course, MetaVal's tool-info module needs to support only a limited set of other tool-info modules and for these it might be fine if version() is never called (it is fine for CPAchecker's module, for example). So you can also check whether this is the case for all tool-info modules that you want to support and keep the current code (though a comment about this would probably be valuable for the future).

I didn't have problems so far, and would deal with them once and if they occur. We can also have a look into this after SV-COMP, if you want to do this just add an issue, assign me and set the due date to beginning of December.